### PR TITLE
Clean deletes, rework indexing

### PR DESCRIPTION
**Pull request by RangerMauve**
- reworked activities indexing code, added author tracking index
- ignore any deletes from authors we haven't interacted with before
- ensure auto-approve follows doesn't bypass moderation
- added more tests to activitystore

### CODE REVIEW

File: src/server/api/inbox.ts [hunk 1]
```diff
@@ -139,8 +139,10 @@ export const inboxRoutes = (cfg: APIConfig, store: Store, apsystem: ActivityPubS
       const submittedActorMention = await apsystem.verifySignedRequest(request, actor)
       to = await apsystem.mentionToActor(submittedActorMention)
     }
+    const replyURL = decodeURIComponent(inReplyTo)
+    request.log.info({ replyURL }, 'fetching replies for post')
```
Comment (fauno) [inline src/server/api/inbox.ts:143]: This is helpful to debug #67 right?

File: src/server/apsystem.ts [hunk 2]
```diff
@@ -361,14 +361,20 @@ export default class ActivityPubSystem {
...
     await actorStore.inbox.add(activity)
 
-    if (activityType === 'Follow' && autoApproveFollow) {
+    if (activityType === 'Follow' && autoApproveFollow && (moderationState !== BLOCKED)) {
       this.log.info({ fromActor, target: activity.object }, 'Auto-approving follow request')
       await this.approveActivity(fromActor, activityId)
     } else if (activityType === 'Undo') {
       await this.performUndo(fromActor, activity)
-    } else if (moderationState === BLOCKED) {
+    } if (moderationState === BLOCKED) {
```
Comment (fauno) [inline src/server/apsystem.ts:377]: is this ok?

File: src/server/apsystem.ts [hunk 3]
```diff
@@ -414,6 +420,8 @@ export default class ActivityPubSystem {
       } else if (typeof activity.object === 'object') {
         // TODO: Account for arrays
         await this.storeObject(fromActor, activity.object as APObject, activity.actor as string)
+      } else {
+        throw new Error(`Unable to load actvity object for ${activityId}.`)
```
Comment (fauno) [inline src/server/apsystem.ts:424]: typo on the error :smile:

File: src/server/store/ActivityStore.ts [hunk 4]
```diff
@@ -1,8 +1,13 @@
 import { AbstractLevel } from 'abstract-level'
-import { APActivity, PublishedField } from 'activitypub-types'
+import { APActivity } from 'activitypub-types'
 import createError from 'http-errors'
 import { Mutex } from 'async-mutex'
 
+export const LATEST_VERSON = '2'
```
Comment (fauno) [inline src/server/store/ActivityStore.ts:6]: typo

File: src/server/store/ActivityStore.ts [hunk 5]
```diff
@@ -32,29 +57,34 @@ export class ActivityStore {
...
-  makeIndexKey (activity: APActivity): string {
-    const publishedAt = getPublished(activity.published)
-    if (typeof activity.id !== 'string') throw createError(400, 'Missing ID in Actvity')
-    return `${publishedAt.toISOString()}-${activity.id}`
-  }
-
   async addToIndex (activity: APActivity): Promise<void> {
-    await this.publishedIndex.put(this.makeIndexKey(activity), activity.id)
+    const { published, id } = activity
+    if (published === undefined || id === undefined) return
```
Comment (fauno) [inline src/server/store/ActivityStore.ts:62]: does this silently ignore activities without a published attribute? are they considered invalid? (i see the `add()` function sets a default date if it missing)

File: src/server/store/ActivityStore.ts [hunk 6]
```diff
@@ -67,29 +97,57 @@ export class ActivityStore {
...
+        // Ensure each activity has a published field
+        for await (const activity of this.db.values()) {
+          if (activity.published === undefined) {
+            activity.published = new Date().toISOString()
+            await this.db.put(this.urlToKey(activity.id), activity)
+          }
+          await this.addToIndex(activity)
+        }
+
+        await this.indexesDB.put('version', LATEST_VERSON)
```
Comment (fauno) [inline src/server/store/ActivityStore.ts:125]: typo (guessing it was autocompleted)

**Review by fauno — CHANGES_REQUESTED** [review]

**Review by akhileshthite — APPROVED** [review]
Thanks! 🚀

File: src/server/api/inbox.ts (hunk 1 above)
Comment (RangerMauve) [reply src/server/api/inbox.ts:143]: Yup!

File: src/server/apsystem.ts (hunk 2 above)
Comment (RangerMauve) [reply src/server/apsystem.ts:377]: good catch

File: src/server/store/ActivityStore.ts (hunk 5 above)
Comment (RangerMauve) [reply src/server/store/ActivityStore.ts:62]: It's mostly to make typescript happy we can safely ignore it

**Pull request merged**